### PR TITLE
AZ-75: Improve AlephBFT output interface

### PR DESCRIPTION
**Pull request by kostekIV**
todo: better docs/comments

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "aleph-bft"
-version = "0.7.0"
+version = "0.8.0"
```
Comment (kostekIV) [inline Cargo.toml:3]: bumped because of the changes in the public api

File: docs/src/aleph_bft_api.md [hunk 2]
```diff
@@ -60,11 +70,11 @@ A typical implementation of KeyBox would be a collection of `N` public keys, an
...
+Consider a blockchain that does not have an intrinsic finality mechanism, so for instance it might be a PoW chain with probabilistic finality or a chain based on PoS that uses some probabilistic or round-robin block proposal mechanism. Each of the `N` nodes in the network has its own view of the chain and at certain moments in time there might be conflicts on what the given nodes consider as the "tip" of the blockchain (because of possible forks or network delays). We would like to design a finality mechanism based on AlephBFT that will allow the nodes to have agreement on what the "tip" is. Towards this end we just need to implement a suitable `DataProvider` object and filtering of network messages for AlephBFT.
```
Comment (timorl) [inline docs/src/aleph_bft_api.md:77]: The description below still uses `send_ordered_batch` rather than the new interface.

File: docs/src/aleph_bft_api.md [hunk 3]
```diff
@@ -142,7 +152,7 @@ def get_data():
 We simply fetch at most 100 transactions from the local pool and return such a list of transactions.
 
 ```
-def send_ordered_batch(batch):
+def data_finalized(batch):
```
Comment (timorl) [inline docs/src/aleph_bft_api.md:155]: Not batch, but a single piece of data. This also requires changes in the code below and some explanation as to when a block should be created at all – no natural batches in this case; probably collect transactions until some timeout or limit is reached?

File: docs/src/aleph_bft_api.md [hunk 4]
```diff
@@ -160,7 +170,7 @@ When it comes to availability, in this case `Data` is not a cryptographic finger
...
+2. For honest nodes that are not "falling behind" significantly (because of network delays or other issues) it is guaranteed that the data items they input in the protocol (from their local `DataProvider` object) will be part of the output stream with a delay of roughly `~round_delay*4` from the time of inputting it. It is hard to define what "falling behind" exactly means, but think of a situation where a node's round `r` unit is always arriving much later then the expected time for round `r` to start. When a node is falling behind from time to time, then there is no issue and its data will be still included in the output stream, however if this problem is chronic, then this node's data might not find its way into the output stream at all. If something like that happens, it most likely means that the `round_delay` is configured too aggresively and one should consider extending the delay.
```
Comment (timorl) [inline docs/src/aleph_bft_api.md:173]: I would rephrase the `output stream` in terms of the new interface to be clearer, i.e. "`data_finalized` will be called on the data items", or something like that.

File: src/lib.rs [hunk 5]
```diff
@@ -35,18 +35,24 @@ pub type SessionId = u64;
...
     fn get_data(&self) -> Data;
-    /// Takes a new ordered batch of data item.
-    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+/// The source of finalization of the units that consensus produces.
+///
+/// The [`FinalizationProvider::data_finalized`] method is called whenever a new unit
+/// (or more precisely the data they carry) is available.
+pub trait FinalizationProvider<Data> {
```
Comment (timorl) [inline src/lib.rs:51]: I'm a bit bothered by the naming – both these traits are called `Provider`, but one provides data to the algorithm, while the other handles pieces of data that have been finalized by it.

Renaming this to `FinalizationHandler` would be better imo, but there might also be an even better naming schema possible here.

File: src/lib.rs [hunk 6]
```diff
@@ -35,18 +35,24 @@ pub type SessionId = u64;
...
     /// Outputs a new data item to be ordered
     fn get_data(&self) -> Data;
-    /// Takes a new ordered batch of data item.
-    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+/// The source of finalization of the units that consensus produces.
+///
+/// The [`FinalizationProvider::data_finalized`] method is called whenever a new unit
+/// (or more precisely the data they carry) is available.
```
Comment (timorl) [inline src/lib.rs:50]: I think this description focuses too much on what happens inside AlephBFT rather than what the user should expect. I would suggest "...is called whenever a piece of data input to the algorithm using [`DataProvider::get_data`] has been finalized, in order of finalization." or something of that nature.

File: src/lib.rs [hunk 7]
```diff
@@ -35,18 +35,24 @@ pub type SessionId = u64;
...
+pub trait FinalizationProvider<Data> {
+    type Error: Debug + 'static;
+
+    /// Data has been finalized. The calls to this function are ordered by the same order as the units.
+    fn data_finalized(&self, data: Data) -> Result<(), Self::Error>;
```
Comment (timorl) [inline src/lib.rs:55]: Is there any specific reason why you want this to return an error? I think in general it's better for the end-user to handle these errors, while here AlephBFT has to do it. Best we can do is log the error, which the user might have done just as well.

**Review by timorl — CHANGES_REQUESTED** [review]
Definitely an improvement over the previous interface, but I had some comments.

File: src/lib.rs (hunk 5 above)
Comment (kostekIV) [reply src/lib.rs:51]: I didnt like this name either. Handler look fine to me, what about `FinalizationNotificator`?

Comment (kostekIV) [reply src/lib.rs:51]: Nevermind I dont like this name anymore :P

File: docs/src/aleph_bft_api.md (hunk 2 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md:76]: fixed

File: docs/src/aleph_bft_api.md (hunk 4 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md]: added

File: src/lib.rs (hunk 5 above)
Comment (kostekIV) [reply src/lib.rs]: Ok, done

File: src/lib.rs (hunk 6 above)
Comment (kostekIV) [reply src/lib.rs]: Done

File: src/lib.rs (hunk 7 above)
Comment (kostekIV) [reply src/lib.rs]: yep, I like it as well. Done

File: docs/src/aleph_bft_api.md (hunk 3 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md]: Since the `Data here is a `Vec` of txs, Im just using it as is (So without collecting and timeout). Let me know if this is fine by you.

File: docs/src/aleph_bft_api.md (hunk 4 above)
Comment (timorl) [reply docs/src/aleph_bft_api.md]: Eugh, this is creating strange jargon now. xd Although technically comprehensible.

How about substituting "will have `FinalizationHandler::data_finalized` called on them with a delay" for "will be part of `FinalizationHandler::data_finalized` with a delay"?

File: src/lib.rs [hunk 8]
```diff
@@ -46,13 +46,12 @@ pub trait DataProvider<Data> {
...
 ///
-/// The [`FinalizationProvider::data_finalized`] method is called whenever a new unit
-/// (or more precisely the data they carry) is available.
-pub trait FinalizationProvider<Data> {
-    type Error: Debug + 'static;
-
-    /// Data has been finalized. The calls to this function are ordered by the same order as the units.
-    fn data_finalized(&self, data: Data) -> Result<(), Self::Error>;
+/// The [`FinalizationHandler::data_finalized`] method is called whenever a piece of data input to the algorithm
+/// using [DataProvider::get_data] has been finalized, in order of finalization.
```
Comment (timorl) [inline src/lib.rs:50]: I think you missed some backticks in the second brackets.

**Review by timorl — APPROVED** [review]
Nice, one minor nit and the medium nit in the unresolved comment from the previous review. Other than that LGTM.

File: src/lib.rs (hunk 8 above)
Comment (kostekIV) [reply src/lib.rs]: Thanks

File: docs/src/aleph_bft_api.md (hunk 4 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md]: done

File: src/lib.rs [hunk 9]
```diff
@@ -35,18 +35,23 @@ pub type SessionId = u64;
...
+}
+
+/// The source of finalization of the units that consensus produces.
+///
+/// The [`FinalizationHandler::data_finalized`] method is called whenever a piece of data input to the algorithm
+/// using [`DataProvider::get_data`] has been finalized, in order of finalization.
+pub trait FinalizationHandler<Data> {
+    /// Data, provided by [DataProvider::get_data], has been finalized.
+    /// The calls to this function follow the order of finalization.
+    fn data_finalized(&self, data: Data);
```
Comment (fixxxedpoint) [inline src/lib.rs:54]: Consider changing to `&mut sefl` so at least tests and examples doesn't need to use `Arc/Mutex` or similar.

File: src/runway.rs [hunk 10]
```diff
@@ -580,7 +592,7 @@ where
 
     async fn on_create(&mut self, u: PreUnit<H>) {
         debug!(target: "AlephBFT-runway", "{:?} On create notification.", self.index());
-        let data = self.data_io.get_data();
+        let data = self.data_provider.get_data();
```
Comment (fixxxedpoint) [inline src/runway.rs:595]: Open question, no need to answer: why it's not an `async` method? Looks a bit scary considering a single threaded executor and some magic that might be happening behind the scene.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]
Just a couple of small requests and questions.

File: docs/src/aleph_bft_api.md [hunk 11]
```diff
@@ -2,19 +2,28 @@
 
 ### 3.1 Required Trait Implementations.
 
-#### 3.1.1 DataIO.
+#### 3.1.1 DataProvider & FinalizationHandler.
 
-The DataIO trait is an abstraction for a component that provides data items and allows to input ordered data items. `DataIO` is parametrized with a `Data` generic type representing the type of items we would like to order. Below we give examples of what these might be.
+The DataProvider trait is an abstraction for a component that provides data items and allows to input ordered data items. `DataProvider` is parametrized with a `Data` generic type representing the type of items we would like to order. Below we give examples of what these might be.
```
[the remark this answers is not in the document]
Comment (kostekIV) [reply docs/src/aleph_bft_api.md]: I have deleted the part with ` and allows to input ordered data items`

File: examples/blockchain/chain/data.rs [hunk 12]
```diff
@@ -1,7 +1,7 @@
 use crate::{chain::BlockNum, network::NetworkData};
-use aleph_bft::OrderedBatch;
-use futures::channel::mpsc::{self, UnboundedReceiver, UnboundedSender};
-use log::debug;
+use futures::channel::{mpsc, mpsc::UnboundedSender};
+use libp2p::futures::channel::mpsc::UnboundedReceiver;
```
[the remark this answers is not in the document]
Comment (kostekIV) [reply examples/blockchain/chain/data.rs]: No, I was relaying to much on auto ide import :|

File: src/lib.rs (hunk 9 above)
Comment (kostekIV) [reply src/lib.rs]: Added

File: docs/src/aleph_bft_api.md [hunk 13]
```diff
@@ -2,19 +2,28 @@
...
+#### 3.1.1 DataProvider & FinalizationHandler.
 
-The DataIO trait is an abstraction for a component that provides data items and allows to input ordered data items. `DataIO` is parametrized with a `Data` generic type representing the type of items we would like to order. Below we give examples of what these might be.
+The DataProvider trait is an abstraction for a component that provides data items. `DataProvider` is parametrized with a `Data` generic type representing the type of items we would like to order. Below we give examples of what these might be.
 
 ```rust
-pub trait DataIO<Data> {
-    type Error: Debug;
+pub trait DataProvider<Data> {
     fn get_data(&self) -> Data;
```
Comment (fixxxedpoint) [inline docs/src/aleph_bft_api.md:11]: Similarly to previous comment: consider changing to `&mut self` so at least tests and examples doesn't need to use `Mutex/Cell` or similar.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]
Small request/suggestion.

File: docs/src/aleph_bft_api.md (hunk 13 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md:11]: added

File: src/runway.rs (hunk 10 above)
Comment (timorl) [reply src/runway.rs:595]: Very good point, it probably should be and this is a good opportunity to change it.

Comment (kostekIV) [reply src/runway.rs]: Added to both `get_data` and `data_finalized`

File: docs/src/aleph_bft_api.md [hunk 14]
```diff
@@ -8,7 +8,7 @@ The DataProvider trait is an abstraction for a component that provides data item
 
 ```rust
 pub trait DataProvider<Data> {
-    fn get_data(&self) -> Data;
+    fn get_data(&mut self) -> Data;
```
Comment (timorl) [inline docs/src/aleph_bft_api.md]: Missing `async`.

**Review by fixxxedpoint — APPROVED** [review]
Great work!

File: docs/src/aleph_bft_api.md (hunk 14 above)
Comment (kostekIV) [reply docs/src/aleph_bft_api.md]: it was added in the last commit
